docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+  which will continue to serve as a pipe between the reactors and the
+  `Router`. The implementation will no longer need the queue
+  implementation and this operation will instead start goroutines that
+  are responsible for routing the messages from the channel to libp2p
+  fundamentals, replacing the current `p2p.Router.routeChannel`.
+
+- The current `p2p.Rotuer.dialPeers` and `p2p.Router.acceptPeers`,
+  which are responsible for establishing outbound and inbound
+  connections respectively, will be removed, along with
+  `p2p.Router.openConnection` and the libp2p connection manager will
+  be responsible for maintaining network connectivity.
+
+- The `p2p.Channel` interface should change to replace the golang
+  channels with a more functional interface for sending messages. The
+  new methods on this object, will take contexts to support safe
+  cancellation and return errors, and generally block rather than
+  return asynchronously. The `Out` channel through which
+  reactors send messages to Peers, will be replaced by a `Send`
+  method, and the Error channel will be replaced by an `Error`
+  method. 
+
+- Rectors will have access to an interface that will allow them to
+  access information from the current state of the Peer information
+  from libp2p. This will supplant the  `p2p.PeerUpdates`
+  subscription.
+
+### Upgrade and Compatibility
+
+Because the routers and all current P2P code is in the `internal`
+package and not part of the public API, the only changes to the public
+API surface area of the tendermint will be different configuration
+file options, replacing the current P2P options with options relevant
+to libp2p.
+
+However, it will not be possible to run a network with both networking
+stacks active at once, so the upgrade to the version of Tendermint
+will need to be coordinated between all nodes of the network.
+
+## Open Questions
+
+- What is the role of Protocol Labs in the implementation of libp2p in
+  tendermint, both during the initial implementation and on an ongoing
+  basis thereafter?
+
+- Should all P2P traffic for a given node be pushed to a single topic,
+  and we assume that a topic maps to a specific ChainID, or should
+  each reactor (or type of message) have its own topic?
+
+- Tendermint presently has a concept of message-type priority, which
+  provides a very course QoS-like functionality and
+  intuitively/theoretically ensures that evidence and consensus
+  messages don't get starved by blocksync/statesync messages. It's
+  unclear if we should attempt to replicate this with libp2p and even
+  if we should.
+
+- Is it possible to attach additional (and potentially arbitrary)
+  information into the DHT as part of the heartbeats between nodes,
+  such as the latest height, and then access that in arbitrary
+  reactors.
+  
+- Does it make sense to have reactors continue to consume inbound
+  messages from a Channel (`In`) or is there another interface or
+  pattern that we should consider?
```

Comment:
A `Recv` analogue of the proposed `Send` method could work well. The reactor could still fan out work in parallel internally if it wants, and in fact a naive implementation could be built around the existing channels.

Reply:
The difficulty with removing channels is that, to work alongside a context, you likely have to have a channel internally anyway so that you can select against `context.Done()`. I somewhat prefer to using a channel in the p2p API so that we can make the contract a little bit more clear. A `Recv() (Value, Error)` method seems pretty clean. It enables a simple loop, while making it clear that the caller has no responsibility to 'close' the `Recv` beyond, perhaps closing the object provided by the p2p code.

Reply:
I don't follow that entirely.  I think it's orthogonal: If the sender and receiver rendezvous, whether or not there is a channel under the covers, they have to agree on what it means to be "finished". If the channels are exposed in the API, then the client has to cooperate to make that happen, but if not, then it can be handled transparently.

Either way, though, we presumably want a context for lifecycle management. It's just that in the case where we hide the channels, the context is the ONLY visible mechanism we really need. But maybe I missed some aspect of your concern?

Reply:
Oh, I think I understand now. I misread what you wrote as advocating _for_ a channel in the API, rather than against. I think we are actually in complete agreement.

Reply:
Sorry!  I accidentally a word.

Reply:
Also agree

Reply:
I don't think that `<-chan` typed channels are super different from `Recv()` methods, though again, I'm open to both ends, and deeply in the camp of "get rid of as many channels as possible.

I think from a code perspective, however, keeping `In` as a Channel rather than replacing it with an `Recv` method makes it easy to write handler functions that consume and process messages from many different channels (e.g. as in the consensus and statesync reactors where we're consuming a number of channels) It's tractable, but that means having a processing goroutine for every channel, which just makes the locking story within the reactor implementation a bit more dodgy.

Reply:
Re: Many input channels. Without reflection, we can only `select` on a static set of inputs, not an arbitrary collection. If we really have cases where, for example, it helps to read exactly k channels at once, I certainly don't oppose using channels, but I haven't really seen an example of that in our code. I think "not exposing channels in the API" is a good default position, that it's fine to deviate from iff we have a concrete need to do so.